File: c++/usaco/bronze/balancing.cpp
#include <bits/stdc++.h>
#define ll long long
using namespace std;

void setIO(string s) {
  ios_base::sync_with_stdio(0); cin.tie(0); 
  freopen((s + ".in").c_str(), "r", stdin);
  freopen((s + ".out").c_str(), "w", stdout);
}

int main() {
  setIO("balancing");
  ll n;
  cin >> n;
  vector<pair<ll, ll>> A(n);
  for (ll i = 0; i < n; ++i) cin >> A[i].first >> A[i].second;
  ll ans = n;
  for (ll i = 0; i < n; ++i) {
    for (ll j = 0; j < n; ++j) {
      ll x = A[i].first + 1, y = A[j].second + 1;
      ll upper_right = 0, lower_right = 0, upper_left = 0, lower_left = 0;
      for (auto& cow: A) {
        if (cow.first > x && cow.second > y) {
          upper_right++;
        } else if (cow.first > x && cow.second < y) {
          lower_right++;
        } else if (cow.first < x && cow.second > y) {
          upper_left++;
        } else {
          lower_left++;
        }
      }
      ans = min(ans, max(max(upper_left, lower_left), max(upper_right, lower_right)));
    }
  }
  cout << ans;
  return 0;
}

File: usaco/bronze/balancing.py
import sys

# sys.stdin = open('input.txt', 'r')
sys.stdin = open('balancing.in', 'r');
sys.stdout = open('balancing.out', 'w')

n, b = [int(x) for x in input().split()]
m = n

grid = [] 
for i in range(n):
  x, y = [int(x) for x in input().split()]
  grid.append((x, y))

for ai in range(n):
  for ci in range(n):
    a = grid[ai][0] - 1
    c = grid[ci][1] - 1
    ne = 0
    es = 0
    sw = 0
    nw = 0
    for x, y in grid:
      if x > a and y > c:
        ne += 1
      if x > a and y < c:
        es += 1
      if x < a and y < c:
        sw += 1
      if x < a and y > c:
        nw += 1
    m = min(m, max(ne, es, sw, nw))

print(m)